File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.py
class Node(object):
    def __init__(self,key,data=None):
        self.key = key
        self.data = data
        self.children = {}

class Trie(object):
    def __init__(self):
        self.head = Node(None)
    def insert(self,string):
        curr_node = self.head
        for char in string:
            if char not in curr_node.children:
                curr_node.children[char] = Node(char)
            curr_node = curr_node.children[char]
        curr_node.data = string

    def starts_with(self,prefix):
        curr_node = self.head
        for char in prefix:
            if curr_node.data:
                return True
            else:
                if char in curr_node.children:
                    curr_node = curr_node.children[char]
                else:
                    return False
        return False
            
    
for _ in range(int(input())):
    n = int(input())
    trie = Trie()
    ans = 'YES'
    arr = []
    for i in range(n):
        tem = input()
        arr.append(tem)
    arr.sort(key=lambda x: len(x))
    for tem in arr:
        if trie.starts_with(tem):
            ans = 'NO'
            break
        else:
            trie.insert(tem)
    print(ans)

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.cc
#include <iostream>
#include <vector>
using namespace std;
char tem[10000][11];

struct Trie
{
	bool finish;
	Trie* Node[10];

	Trie()
	{
		finish = false;
		for (int i = 0; i < 10; i++) {
			Node[i] = NULL;
		}
	}

	~Trie() {
		for (int i = 0; i < 26; i++) {
			if (Node[i]) {
				delete Node[i];
			}
		}
	}

	void insert(const char* key) {
		if (*key == '\0') {
			finish = true;
			return;
		}
		else {
			int curr = *key - '0';
			if (Node[curr] == NULL) {
				Node[curr] = new Trie();
			}
			Node[curr]->insert(key + 1);
		}
	}
	bool find(const char* key) {
		if (*key == '\0') {
			return false;
		}
		if (finish == true) {
			return true;
		}
		int curr = *key - '0';
		if (Node[curr] == NULL) return false;
		return Node[curr] -> find(key + 1);
	}
};
int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	int n,t;
	cin >> n;
	for (int j = 0; j < n; j++) {
		cin >> t;
		Trie* trie = new Trie();
		string ans = "YES";
		for (int i = 0; i < t; i++) {
			cin >> tem[i];
			trie->insert(tem[i]);
		}
		int i;
		for (i = 0; i < t; i++) {
			if (trie->find(tem[i])) {
				cout << "NO" << "\n";
				break;
			}
		}
		if (i == t) {
			cout << "YES" << "\n";
		}
		
	}
	
	return 0;
	
}